crates/wasi-threads/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+//! Implement [`wasi-threads`].
+//!
+//! [`wasi-threads`]: https://github.com/WebAssembly/wasi-threads
+
+use anyhow::{anyhow, Result};
+use rand::Rng;
+use std::thread;
+use wasmtime::{Caller, Linker, Module, SharedMemory, Store};
+use wasmtime_wasi::maybe_exit_on_error;
+
+// This name is a function export designated by the wasi-threads specification:
+// https://github.com/WebAssembly/wasi-threads/#detailed-design-discussion
+const WASI_ENTRY_POINT: &str = "wasi_thread_start";
+
+pub struct WasiThreadsCtx<T> {
+    module: Module,
+    linker: Linker<T>,
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Linker<T>) -> Self {
+        Self { module, linker }
+    }
```

Comment:
There are two unwraps below during the `spawn` operation's closure: 

* Using `Linker::instantiate`
* The call to `get_typed_func`

Both of these, however, can be type-checked ahead of time at this location I believe by inspecting `Module` and `Linker<T>`. Could you either add a TODO to perform such type-checking here or implement it in this PR?